### PR TITLE
Feature/hasura allow operations in

### PR DESCRIPTION
**Pull request by rhyslbw**
Adds a standalone utility `hasura-allow-operations-in`

> Populate the [Hasura allow-list](https://hasura.io/docs/1.0/graphql/manual/deployment/allow-list.html) from found GraphQL operations in a path, including queries, mutations, and subscriptions.

... and implements it within a basic Node.js program for Cardano GraphQL clients to lock their own instance down by setting `HASURA_URI` and passing a glob pattern leading to `.graphql` files. The primary reason for this additional layer is to include operations used by Cardano GraphQL without leaking internal concerns. It might be best to publish this in the cli package too, statically including the queries without needing the source repo, but for now this will prove the design.

```
HASURA_URI=http://localhost:8090 yarn hasura:allow-operations-in 'PATH/TO/**/*.graphql'
```

:memo: [README](https://github.com/input-output-hk/cardano-graphql/blob/feature/hasura-allow-operations-in/hasura/hasura-allow-operations-in/README.md)

:warning: [Breaking change with config](https://github.com/input-output-hk/cardano-graphql/commit/6bdf99651a729f1a5b978e8e6d4f8b806e149f5d). `HASURA_URI` :point_right:  `http://host.8080`

### CODE REVIEW

**Comment by rhyslbw** [comment]
@KtorZ It's a larger diff due to a couple of required refactors, but the commits are clean and isolate the changes well.

File: nix/packages.nix [hunk 1]
```diff
@@ -7,7 +7,33 @@
...
+          lib.hasPrefix "/node_modules" sansPrefix ||
+          lib.hasPrefix "/build" sansPrefix;
+        in_whitelist =
+          (type == "directory") ||
+          (lib.hasSuffix ".yml" name) ||
+          #(lib.hasSuffix ".js" name) ||
+          (lib.hasSuffix ".ts" name) ||
+          #(lib.hasSuffix ".tsx" name) ||
+          (lib.hasSuffix ".json" name) ||
+          #(lib.hasSuffix ".graphql" name) ||
```
Comment (rhyslbw) [inline nix/packages.nix:24]: @disassembler you will need `.graphql` files for the build

File: nix/packages.nix [hunk 2]
```diff
@@ -7,7 +7,33 @@
...
+          #(lib.hasPrefix "/source/public/assets" sansPrefix) ||
+          (lib.hasPrefix "/source" sansPrefix) ||
+          baseName == ".babelrc" ||
+          baseName == "package.json" ||
+          baseName == "next.config.js" ||
```
Comment (rhyslbw) [inline nix/packages.nix:29]: This project doesn't use `Next.js`

File: nix/packages.nix [hunk 3]
```diff
@@ -7,7 +7,33 @@
...
+          (type == "directory") ||
+          (lib.hasSuffix ".yml" name) ||
+          #(lib.hasSuffix ".js" name) ||
+          (lib.hasSuffix ".ts" name) ||
+          #(lib.hasSuffix ".tsx" name) ||
+          (lib.hasSuffix ".json" name) ||
+          #(lib.hasSuffix ".graphql" name) ||
+          #(lib.hasPrefix "/source/public/assets" sansPrefix) ||
+          (lib.hasPrefix "/source" sansPrefix) ||
+          baseName == ".babelrc" ||
```
Comment (rhyslbw) [inline nix/packages.nix:27]: This project doesn't use Babel

File: nix/packages.nix [hunk 4]
```diff
@@ -7,7 +7,33 @@
...
+        in_whitelist =
+          (type == "directory") ||
+          (lib.hasSuffix ".yml" name) ||
+          #(lib.hasSuffix ".js" name) ||
+          (lib.hasSuffix ".ts" name) ||
+          #(lib.hasSuffix ".tsx" name) ||
+          (lib.hasSuffix ".json" name) ||
+          #(lib.hasSuffix ".graphql" name) ||
+          #(lib.hasPrefix "/source/public/assets" sansPrefix) ||
+          (lib.hasPrefix "/source" sansPrefix) ||
```
Comment (rhyslbw) [inline nix/packages.nix:26]: Directory is `/src`

File: hasura/hasura-allow-operations-in/src/cli.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+const hasuraUri = process.argv[3];
+const allowIntrospection = process.argv[4] === 'true';
+
+if (sourcePath === undefined) {
+  throw new Error('Source path must be passed as first argument');
+}
+
+if (hasuraUri === undefined) {
+  throw new Error('Hasura URI must be passed as the second argument');
+}
```
Comment (KtorZ) [inline hasura/hasura-allow-operations-in/src/cli.ts:14]: Any reason why this one is passed as an argument here whereas it's passed as an ENV var everywhere else?

Comment (rhyslbw) [reply hasura/hasura-allow-operations-in/src/cli.ts:14]: I was going to annotate this to explain. I've implemented a standalone utility that is useful outside the context of Cardano GraphQL, and possibly even worth lifting to another repo (can handle that later though). In this context all arguments are provided in the command.

The ENV is an implementation detail from Cardano GraphQL, and is mapped through when the [package is implemented](https://github.com/input-output-hk/cardano-graphql/blob/feature/hasura-allow-operations-in/src/util/hasuraAllowOperationsIn.ts#L21)

File: src/util/hasuraAllowOperationsIn.ts [hunk 6]
```diff
@@ -13,11 +13,22 @@ if (process.env.HASURA_URI === undefined) {
...
+    ({
+      introspectionAllowed,
+      operationDefinitionsFound,
+      addedCount,
+      existingCount
+    }) => {
+      console.log(
+        `Introspection allowed: ${introspectionAllowed} | Found: ${operationDefinitionsFound.length} | Added: ${addedCount} | Existing: ${existingCount}`
+      )
+      if (process.env.DEBUG) {
```
Comment (rhyslbw) [inline src/util/hasuraAllowOperationsIn.ts:27]: @disassembler `DEBUG=true` to see which files are found

**Comment by rhyslbw** [comment]
Closing in favour of #166  #167 #168

**Pull request closed without merging**